File: examples/csharp/Scripts/ListEffectScene.cs
using Godot;

namespace FairyGUI.Examples;

public partial class ListEffectScene : DemoSceneBase
{
    private GList _list;

    protected override void ContinueInit()
    {
        UIPackage.AddPackage("res://Resources/UI/Extension");
        _view = UIPackage.CreateObject("Extension", "Main");
        _groot.AddChild(_view);

        _list = ((GComponent)_view).GetChild("mailList") as GList;
        if (_list == null)
        {
            GD.PushError("ListEffectScene: mailList is missing or not a GList.");
            return;
        }

        CallDeferred(MethodName.PopulateList);
    }

    private void PopulateList()
    {
        if (!IsUiActive() || _list == null)
            return;

        _list.RemoveChildrenToPool();

        for (var i = 0; i < 10; i++)
        {
            var item = _list.AddItemFromPool();
            if (item == null)
                continue;

            var timeText = item.GetChild("timeText");
            timeText?.SetText("5 Nov 2015 16:24:33");

            var readCtrl = item.GetController("IsRead");
            readCtrl?.SetSelectedIndex(i % 2 == 0 ? 0 : 1);

            var fetchCtrl = item.GetController("c1");
            fetchCtrl?.SetSelectedIndex(i % 3 != 0 ? 0 : 1);

            item.SetTitle("Mail title here");
            // 与 Unity MailItem.PlayEffect 一致：配置完立刻隐藏，避免先显示再播 transition
            item.SetVisible(false);
        }

        _list.EnsureBoundsCorrect();

        var delay = 0.0f;
        for (var i = 0; i < 10; i++)
        {
            var item = _list.GetChildAt(i);
            if (item == null)
                break;
            if (!_list.IsChildInView(item))
                break;

            ((GComponent)item).GetTransition("t0")?.Play(1, delay);
            delay += 0.2f;
        }
    }
}

File: examples/ts/Scripts/ListEffectScene.ts
/// <reference path="../fairygui.d.ts" />

import { Callable } from "godot";
import { DemoSceneBase } from "./DemoSceneBase";

export default class ListEffectScene extends DemoSceneBase {
    private _list: GList | null = null;

    protected ContinueInit(): void {
        UIPackage.addPackage("res://Resources/UI/Extension");
        this._view = UIPackage.createObject("Extension", "Main");
        this._groot!.addChild(this._view);

        this._list = this._view.getChild("mailList") as GList | null;
        if (this._list == null) {
            return;
        }

        for (let i = 0; i < 10; i++) {
            const item = this._list.addItemFromPool() as GComponent | null;
            if (item == null) {
                continue;
            }
            const timeText = item.getChild("timeText");
            if (timeText != null) {
                timeText.setText("5 Nov 2015 16:24:33");
            }

            const readCtrl = item.getController("IsRead");
            if (readCtrl != null) {
                readCtrl.selectedIndex = i % 2 === 0 ? 0 : 1;
            }

            const fetchCtrl = item.getController("c1");
            if (fetchCtrl != null) {
                fetchCtrl.selectedIndex = i % 3 !== 0 ? 0 : 1;
            }

            item.setTitle("Mail title here");
        }

        this._list.ensureBoundsCorrect();
        let delay = 1.0;
        for (let i = 0; i < 10; i++) {
            const item = this._list.getChildAt(i) as GComponent | null;
            if (item == null) {
                break;
            }
            if (this._list.isChildInView(item)) {
                const trans = item.getTransition("t0");
                if (trans != null) {
                    trans.play(1, delay);
                }
                delay += 0.2;
            } else {
                break;
            }
        }
    }
}
